styx-api-service/src/main/java/com/spotify/styx/api/ServiceAccountUsageAuthorizer.java

```diff
@@ -0,0 +1,445 @@
+
+              // Check if the principal has been granted the service account user role on the SA itself
+              () -> serviceAccountPolicyAccess(serviceAccount, principalEmail)
+                  .map(type -> String.format("Principal %s has role %s on service account %s %s",
+                      principalEmail, serviceAccountUserRole, serviceAccount, type))));
+        } catch (ResponseException e) {
+          return Either.left(e.getResponse());
+        }
+      });
+
+      // Propagate response exception
+      if (result.isLeft()) {
+        throw new ResponseException(result.left().get());
+      }
+
+      final Optional<String> accessMessage = result.right().get();
+
+      // Grant access?
+      if (accessMessage.isPresent()) {
+        logAuthorization(workflowId, serviceAccount, enforce, accessMessage.get(), cached.get());
+        return;
+      }
+
+      // Deny access?
+      logDenial(workflowId, serviceAccount, enforce, principalEmail, cached.get());
+      if (enforce) {
+        throw denialResponseException(serviceAccount, principalEmail, projectId);
+      }
+    }
+
+    private static <K, V> V get(Cache<K, V> cache, K key, Callable<V> loader) {
+      try {
+        return cache.get(key, loader);
+      } catch (ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private ResponseException denialResponseException(String serviceAccount, String principalEmail, String projectId) {
+      return new ResponseException(Response.forStatus(
+          FORBIDDEN.withReasonPhrase("The user " + principalEmail + " must have the role " + serviceAccountUserRole
+              + " on the project " + projectId + " or the service account " + serviceAccount +
+              ", either through a group membership (recommended) or directly")));
+    }
+
+    private void logDenial(WorkflowId workflowId, String serviceAccount, boolean enforce, String principalEmail,
+        boolean cached) {
+      log.info("[AUDIT] Principal {} denied use of service account {} in workflow {} (Enforcing: {}, Cache: {})",
+          principalEmail, serviceAccount, workflowId.toKey(), enforce, cached ? CACHE_HIT : CACHE_MISS);
+    }
+
+    private void logAuthorization(WorkflowId workflowId, String serviceAccount, boolean enforce, String accessMessage,
+        boolean cached) {
+      log.info("[AUDIT] {}, authorizing use of service account {} in workflow {} (Enforcing: {}, Cache: {})",
+          accessMessage, serviceAccount, workflowId.toKey(), enforce, cached ? CACHE_HIT : CACHE_MISS);
+    }
+
+    private String serviceAccountProjectId(String serviceAccount) {
+      final Matcher matcher = USER_CREATED_SERVICE_ACCOUNT_PATTERN.matcher(serviceAccount);
+      if (!matcher.matches()) {
+        throw new ResponseException(Response.forStatus(
+            BAD_REQUEST.withReasonPhrase("Not a user created service account: " + serviceAccount)));
```

Comment:
`Authenticator` allows for example GCE default service account. Shall we allow that as well here?

Reply:
Yeah, iirc we can't even create keys for gcp managed service accounts (like the default GCE account).